### ISSUE TITLE
Work Functions are not provided with the correct block hash input

### ISSUE THREAD

**Issue by GoudaSoftware** [stjet/banani#8]
I tried creating a custom work function but discovered the functions are provided with a block hash of the block that is intending to be published. It should actually be the [frontier block or the public key](https://docs.nano.org/integration-guides/work-generation/#work-equation) of the account attempting to publish.

For example, I believe [this](https://github.com/stjet/banani/blob/53482cb9fa3594279a2901179048f760c99e7174/wallet.ts#L92) should be `info.frontier` instead of `s_block_hash`. I suppose receives are slightly more complicated due to the potential absence of a frontier.

---

If it's helpful to see, this is the work function code I am using. This is what bananojs does, except using the slightly different blake2b dependency of this project. Note the `blockHashOverride`, which I needed to get this successfully working by providing frontier block hashes manually copied from creeper.

```
import {BlockHash, hex_to_uint8array, uint8array_to_hex, WorkFunction} from "banani";
import blake2b from "blake2b";

// TODO is this the banano's or nano's current min? Can it be smaller for receive operations?
const WORK_MIN: bigint = BigInt('0xfffffe0000000000'); 

function isWorkValid(hashBytes: Uint8Array, workBytes: Uint8Array): boolean {
  const calculated: Uint8Array = blake2b(8, null, null, null, true)
    .update(workBytes)
    .update(hashBytes)
    .digest().reverse();
  const calculatedHex: string = uint8array_to_hex(calculated);
  const calculatedBigInt: bigint = BigInt("0x" + calculatedHex);

  const isValid: boolean = calculatedBigInt >= WORK_MIN;
  if (isValid) {
    console.log(`Found valid work ${calculatedBigInt} >= ${WORK_MIN} from work bytes ${workBytes}`);
  }
  return isValid;
}

function incrementBytes(bytes: Uint8Array) {
  let x = bytes.length - 1;
  for (x; x >= 0; x--) {
    if (bytes[x] ^ 0xff) {
      bytes[x]++;
      return;
    } else {
      bytes[x] = 0;
    }
  }
}

async function doHashCpuWork(hash: BlockHash, workBytes: Uint8Array): Promise<string> {
  console.log('STARTED doHashCpuWork', hash, uint8array_to_hex(workBytes));
  const hashBytes = hex_to_uint8array(hash);

  let isValid = isWorkValid(hashBytes, workBytes);
  while (!isValid) {
    incrementBytes(workBytes);
    isValid = isWorkValid(hashBytes, workBytes);
  }
  const pow = uint8array_to_hex(workBytes.reverse());
  console.log(`POW: ${pow}`);
  return pow;
}

export const cpuWorkFunction: WorkFunction = async function(blockHash: BlockHash): Promise<string> {
  const blockHashOverride = "...";
  const workBytes = new Uint8Array(8);
  return doHashCpuWork(blockHashOverride, workBytes);
}
```

I'm setting this like `wallet.work_function = cpuWorkFunction;`.

**Comment by stjet** [comment]
Ha, that's hilarious. I did the same mistake with [bananopie](https://github.com/stjet/bananopie/issues/5). Will fix

**Comment by stjet** [comment]
Unrelated, is there any reason you do `wallet.work_function = cpuWorkFunction;`?

Because the constructor for `Wallet` should accept the `work_function` parameter and set it as a property:

```typescript
  constructor(rpc: RPCInterface, seed: string, index: number = 0, work_function?: WorkFunction) {
    this.rpc = rpc;
    if (typeof seed !== "string" || seed?.length !== 64) throw Error("Seed needs to be 64 character (hex) string");
    this.seed = seed;
    this.index = index;
    this.work_function = work_function;
  }
```

Does that not work?

**Comment by stjet** [comment]
@GoudaSoftware published v1.0.6 (commit f185cad), does that fix the issue?

**Comment by GoudaSoftware** [comment]
@stjet Just tested out the new version and it worked for sends and receives! Thank you! And regarding the constructor, that works fine. I just didn't notice that before. Thanks for pointing it out.
